### PR TITLE
Add Unit Test for Genformer Training Script

### PR DESCRIPTION
**Pull request by nvmvle**
### Description
This PR refactors the Geneformer training tests to improve efficiency and maintainability:

1. **Moved TensorBoard verification utility to a shared location**: The `verify_tensorboard_logs` function has been moved from the test file to a new module `bionemo.testing.tensorboard`, making it reusable across different test suites.

2. **Updated metric naming**: Changed the outdated metric name `train_tflops_per_sec_per_gpu` to the current name `TFLOPS_per_GPU` to align with the latest naming conventions.

3. **Optimized test fixture scope**: Changed the `data_path` fixture scope from function-level to module-level to reduce redundant data loading operations across tests, improving test execution time.

4. **Enabled TensorBoard and TFLOPS callbacks by default**: Set `create_tensorboard_logger` and `create_tflops_callback` to `True` by default in the parametrized tests to ensure comprehensive metric logging during testing.

### Type of changes

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [x]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

* If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
* If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
# The verify_tensorboard_logs function can now be imported from bionemo.testing
from bionemo.testing.tensorboard import verify_tensorboard_logs, TENSORBOARD_AVAILABLE

# Use in any test that needs to verify TensorBoard logs
def test_my_training_with_tensorboard(tmpdir):
    # ... run training that produces TensorBoard logs ...
    
    tb_log_dir = Path(tmpdir) / "tb_logs"
    expected_metrics = ["lr", "TFLOPS_perGPU", "train_step_timing"]
    
    # Verify logs contain expected metrics
    verify_tensorboard_logs(tb_log_dir, expected_metrics, min_steps=5)
```

### Pre-submit Checklist

 - [x] I have tested these changes locally
 - [ ] I have updated the documentation accordingly
 - [x] I have added/updated tests as needed
 - [x] All existing tests pass successfully

## Summary by CodeRabbit

* **Tests**
  * Expanded end-to-end tests to validate TensorBoard logs, TFLOPS and other training metrics, checkpoint presence/absence, and experiment/run directory structure; added stricter checks for metric progression and clearer failure messages.
  * Updated pretraining test to run the CLI with TensorBoard and performance logging enabled and validate resulting event files.

* **Chores**
  * Added a reusable utility to verify TensorBoard event files and expected metrics across steps.

### CODE REVIEW

File: sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py [hunk 1]
```diff
@@ -44,7 +51,15 @@ def test_bionemo2_rootdir(data_path):
 
 
 @pytest.mark.parametrize("create_checkpoint_callback", [True, False])
-def test_main_runs(tmpdir, create_checkpoint_callback: bool, data_path: Path):
+@pytest.mark.parametrize("create_tensorboard_logger", [True])
```
Comment (dorotat-nv) [inline sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py]: test_main_... tests are very time consuming. Ideally, would be great to have only one call that checks everything, so set in `test_main_runs`

` create_tensorboard_logger=True, create_tflops_callback=True`  and add there checks from `test_tensorboard_metrics_and_directory_structure`

File: sub-packages/bionemo-testing/src/bionemo/testing/tensorboard.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pathlib import Path
+
+import pytest
+
+
+try:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-testing/src/bionemo/testing/tensorboard.py]: we dont need this flag. Tensorboard is one of the dependencies in the environment.

File: sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py [hunk 3]
```diff
@@ -56,49 +70,90 @@ def test_main_runs(tmpdir, create_checkpoint_callback: bool, data_path: Path):
             result_dir=result_dir,
             wandb_project=None,
             wandb_offline=True,
-            num_steps=5,
-            limit_val_batches=1,
-            val_check_interval=2,
+            num_steps=20,  # Run for more steps to ensure metrics are logged
```
Comment (dorotat-nv) [inline sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py]: can we change it back to smaller number of step? why this change? and lets limit val batches to 1

File: sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py [hunk 4]
```diff
@@ -22,14 +22,17 @@
 import pytest
 from lightning.fabric.plugins.environments.lightning import find_free_network_port
 
+
 from bionemo.core.data.load import load
 from bionemo.geneformer.scripts.train_geneformer import get_parser, main
 from bionemo.llm.model.biobert.transformer_specs import BiobertSpecOption
 from bionemo.llm.utils.datamodule_utils import parse_kwargs_to_arglist
 from bionemo.testing import megatron_parallel_state_utils
+from bionemo.testing.tensorboard import verify_tensorboard_logs
```
Comment (dorotat-nv) [inline sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py]: it seems this is not used

**Comment by nvmvle** [comment]
/ok to test https://github.com/NVIDIA/bionemo-framework/pull/943/commits/30fc673d5d3ad6ffb31a10f29381673b7eaa7d2d

File: sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py [hunk 5]
```diff
@@ -306,3 +360,64 @@ def test_limit_val_batches_is_int(required_args_reference, limit_val_batches):
     arglist = parse_kwargs_to_arglist(required_args_reference)
     parser = get_parser()
     parser.parse_args(arglist)
+
+
+
+
+
+@pytest.mark.slow
+def test_pretrain_cli_with_tensorboard(tmpdir, data_path):
```
Comment (dorotat-nv) [inline sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py]: similarly for this test, lets merge it with test_pretrain_cli - please add the tensorboard parameters and test if it is executed

File: sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py [hunk 6]
```diff
@@ -306,3 +360,64 @@ def test_limit_val_batches_is_int(required_args_reference, limit_val_batches):
     arglist = parse_kwargs_to_arglist(required_args_reference)
     parser = get_parser()
     parser.parse_args(arglist)
+
```
Comment (dorotat-nv) [inline sub-packages/bionemo-geneformer/tests/bionemo/geneformer/scripts/test_train_geneformer.py]: code format checker probably will not like these empty lines, please install pre commit

**Review by dorotat-nv — CHANGES_REQUESTED** [review]

**Comment by nvmvle** [comment]
/ok to test https://github.com/NVIDIA/bionemo-framework/pull/943/commits/dc6bf14172deb7c47b583e7f7e4c4bd374bf0914

**Review by malcolmgreaves — APPROVED** [review]

**Review by dorotat-nv — APPROVED** [review]

File: sub-packages/bionemo-testing/src/bionemo/testing/tensorboard.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    event_acc.Reload()
+
+    # Get available scalar tags
+    scalar_tags = event_acc.Tags()["scalars"]
+
+    # Check that expected metrics are present
+    for metric in expected_metrics:
+        # Check if metric exists in any form (might have prefixes like "train/" or suffixes)
+        metric_found = any(metric in tag for tag in scalar_tags)
+        assert metric_found, f"Expected metric '{metric}' not found in TensorBoard logs. Available tags: {scalar_tags}"
```
Comment (dorotat-nv) [inline sub-packages/bionemo-testing/src/bionemo/testing/tensorboard.py]: assert should be used only in unit tests, please return error

File: sub-packages/bionemo-testing/src/bionemo/testing/tensorboard.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    # Verify we have logged data for at least min_steps
+    if scalar_tags:
+        # Get the first available metric to check step count
+        first_metric = scalar_tags[0]
+        events = event_acc.Scalars(first_metric)
+        assert len(events) >= min_steps, f"Expected at least {min_steps} steps logged, but found {len(events)}"
```
Comment (dorotat-nv) [inline sub-packages/bionemo-testing/src/bionemo/testing/tensorboard.py]: as above

**Review by dorotat-nv — CHANGES_REQUESTED** [review]
please fix format issues https://github.com/NVIDIA/bionemo-framework/actions/runs/17731355920/job/50383039749?pr=943

**Review by dorotat-nv — APPROVED** [review]
